infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogCompactor.java

```diff
@@ -0,0 +1,410 @@
+
+            log.debug("Started compacting garbage log segment: {}", inputStreamSegment.getFilePath());
+            startTime = System.currentTimeMillis();
+            // This lock is need as garbage segment can be concurrently written while compacting.
+            // But the lock should not take long time as garbage segment is small (several MB).
+            try (AutoCloseableLock ignored = GarbageLogSegment.getSegmentLock().acquireWriteLock(ordinal)) {
+                compactGarbageSegment(inputGarbageSegment, outputGarbageSegment,
+                        compactionFeedback.garbageEntriesToPrune);
+                segmentManager.remapCompactedSegment(inputGarbageSegment, outputGarbageSegment);
+            }
+
+            // Close input and output garbage segment channels.
+            closeSegments(inputGarbageSegment, outputGarbageSegment);
+            span = System.currentTimeMillis() - startTime;
+            log.debug("Compacted garbage log segment: {}, took {} ms", inputGarbageSegment.getFilePath(), span);
+
+        } catch (Throwable t) {
+            log.error("compactSegment: encountered an exception, " +
+                    "compaction on segment: {} might be uncompleted.", ordinal, t);
+            closeSegments(inputStreamSegment, outputStreamSegment, inputGarbageSegment, outputGarbageSegment);
+            throw t;
+        }
+    }
+
+    private CompactionFeedback compactStreamSegment(StreamLogSegment inputStreamSegment,
+                                                    StreamLogSegment outputStreamSegment,
+                                                    GarbageLogSegment inputGarbageSegment) {
+        int flushBatchCount = 0;
+        List<LogData> batch = new ArrayList<>();
+        Map<Long, SMRGarbageEntry> garbageInfoMap = inputGarbageSegment.getGarbageEntryMap();
+        CompactionFeedback compactionFeedback = new CompactionFeedback();
+
+        // Scan the input StreamLogSegment and compact entries.
+        for (IndexedLogEntry indexedEntry : inputStreamSegment) {
+            LogData logData = getLogData(indexedEntry.logEntry);
+            long address = logData.getGlobalAddress();
+            LogData newLogData = compactStreamLogEntry(logData, address,
+                    garbageInfoMap.get(address), compactionFeedback);
+            // If the entire LogData is compacted, do not append to new segment.
+            if (newLogData != null) {
+                batch.add(newLogData);
+            }
+
+            // Write batch to the output segment.
+            flushBatchCount = writeBatchToCompactionOutput(batch,
+                    STREAM_COMPACTION_BATCH_SIZE, flushBatchCount, outputStreamSegment);
+        }
+        writeBatchToCompactionOutput(batch, outputStreamSegment);
+
+        return compactionFeedback;
+    }
+
+    @Nullable
+    private LogData compactStreamLogEntry(LogData logData, long address,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'compactStreamLogEntry(LogData, long, SMRGarbageEntry, CompactionFeedback)' has an NPath complexity of 928](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)